keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+
+### Context
+
+We assume the following context of OpenAPI schemas in the Kubernetes architecture and ecosystem:
+
+#### Consumers of the restricted OpenAPI Schema
+
+* server side validation (consumes structural schema including value validation)
+* OpenAPI publishing
+  * client side validation
+  * kubectl explain
+  * client generators (https://github.com/kubernetes-client)
+  * doc generator
+  * kube-aggregator
+  * IDEs like IntelliJ for input completion
+* server side apply
+* pruning
+* defaulting
+* possibly in the future:
+  * server side read-only update validation
+  * Protobuf publishing
+
+#### Producers of the restricted OpenAPI Schema
+
+* kube-openapi's **openapi-gen** mechanically producing go-openapi definitions for native types compiled into kube-apiserver
+* **crd-gen/crd-schema-gen** in https://github.com/kubernetes-sigs/controller-tools mechanically producing CRD validation schemas from Golang types with tags
+* a **developer** adding custom OpenAPI schemas for openapi-gen or crd-gen/crd-schema-gen.
+* a **developer** manually writing complete CRD validation schemas.
+
+![Information flow of OpenAPI schemas in Kubernetes](20190425-structural-openapi-context.png)
+
+### Criteria
+
+To allow only a subset of features of OpenAPI v3, we need to have guidelines for what features we want to restrict and what features we need to permit. We would like to select a single canonical way in OpenAPI to represent these Kubernetes-like types:
+
+* atomic types: primitive types, maps, arrays.
+
+  Examples:
+  * string maps like labels and annotations, but also `map[string]int64`
+  * IPv4 vs IPv6 strings: same fields with different formats
+  * embedded PodSpec
+  
+* discriminator union types (as described in [KEP: Union Types](https://github.com/kubernetes/enhancements/pull/926)).
+  
+  Examples:
+  * VolumeSource, DeploymentStrategy, …
+
+* a very restrictive, fixed set of polymorphic types.
+ 
+  Examples:
+  * IntOrString
+  * RawExtensions, i.e. unconstrained, embedded toplevel types
+  * arbitrary JSON
+  
+Note: we explicitly accept that polymorphic types (like IntOrString) are an anti-pattern for Kubernetes-like types and we do not want to encourage its use beyond IntOrString.
+
+As a soft criterion to validate the direction, we want that restricted CRD OpenAPI schemas represent a model that can also be represented with other parts of the system: 
+
+    Golang types <=> OpenAPI <=> Protobuf.
+
+Each of these is to be restricted to lead to a canonical mapping.
+
+## Proposal
+
+We propose to define a subset of OpenAPI v3 called _structural schema_ following this guiding principle:
+
+**Structural Completeness**: all types and the possible object fields within a JSON object must be apparent from a structural schema without understanding the logical junctors `anyOf`, `allOf`, `oneOf` and `not`.
+
+Example:
```

Comment:
make it clearer up front this is an example of what we want to avoid

Reply:
done